app/models/voxelytics/VoxelyticsService.scala

```diff
@@ -0,0 +1,205 @@
+                                endTime: Instant,
+                                memory: StatisticsEntry,
+                                cpuUser: StatisticsEntry,
+                                cpuSystem: StatisticsEntry,
+                                duration: StatisticsEntry)
+
+object ChunkStatisticsEntry {
+  implicit val jsonFormat: OFormat[ChunkStatisticsEntry] = Json.format[ChunkStatisticsEntry]
+}
+
+case class ArtifactChecksumEntry(taskName: String,
+                                 artifactName: String,
+                                 path: String,
+                                 resolvedPath: String,
+                                 timestamp: Instant,
+                                 checksumMethod: String,
+                                 checksum: String,
+                                 fileSize: Long,
+                                 lastModified: Instant)
+
+object ArtifactChecksumEntry {
+  implicit val jsonFormat: OFormat[ArtifactChecksumEntry] = Json.format[ArtifactChecksumEntry]
+}
+
+class VoxelyticsService @Inject()(voxelyticsDAO: VoxelyticsDAO)(implicit ec: ExecutionContext) extends FoxImplicits {
+
+  def checkAuth(runId: ObjectId, user: User): Fox[Unit] =
+    for {
+      runUserId <- voxelyticsDAO.getUserIdForRun(runId)
+    } yield bool2Fox(user.isAdmin || runUserId == user._id)
+
+  def checkAuthForWorkflowCreation(runName: String, user: User): Fox[Unit] =
+    for {
+      runUserId <- voxelyticsDAO.getUserIdForRunOpt(runName, user._organization)
+    } yield bool2Fox(user.isAdmin || runUserId.forall(_ == user._id))
+
+  def runPublicWrites(run: RunEntry, tasks: List[TaskRunEntry]): JsObject =
+    Json.toJson(run).as[JsObject] ++ Json.obj(
+      "tasks" -> tasks.map(Json.toJson(_))
+    )
+
+  private def artifactEntryPublicWrites(artifact: ArtifactEntry): (String, JsObject) =
+    (artifact.name,
+     artifact.metadata ++
+       Json.obj(
+         "artifactId" -> artifact.artifactId.id,
+         "taskId" -> artifact.taskId.id,
+         "name" -> artifact.name,
+         "path" -> artifact.path,
+         "fileSize" -> artifact.fileSize,
+         "inodeCount" -> artifact.inodeCount,
+         "version" -> artifact.version,
+         "taskName" -> artifact.taskName
+       ))
+
+  def artifactsPublicWrites(artifacts: List[ArtifactEntry]): JsObject = {
+    val artifactsByTask = artifacts.groupBy(_.taskName)
+    JsObject(artifactsByTask.map(artifactKV => {
+      val taskName = artifactKV._1
+      val artifacts = artifactKV._2
+      (taskName, JsObject(artifacts.map(artifactEntryPublicWrites)))
+    }))
+  }
+
+  def workflowConfigPublicWrites(workflowConfig: JsObject, tasks: List[TaskEntry]): JsObject =
+    workflowConfig ++ Json.obj("tasks" -> JsObject(tasks.map(t => (t.name, t.config ++ Json.obj("task" -> t.task)))))
+
+  def aggregateBeginEndTime(runs: List[RunEntry]): (VoxelyticsRunState, Instant, Option[Instant]) = {
+    val state = runs.maxBy(_.beginTime).state
```

Comment:
maxBy throws an exception when called on an empty list. What is the expected behavior in this case?

Reply:
alright :) maybe add a comment in this line that the callers take care of this